### ISSUE TITLE
TypeVar isn't preserved for bound object

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#1192]
The following code should not produce an error but it does.

```python
from typing import TypeVar

TP = TypeVar("TP", bound="P")

class P:
    def chain(self: TP) -> TP:
        ...

def fn(p2: TP) -> TP:
    p2 = p2.chain()  # error
    return p2
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.88, which I just published. It will also be included in the next release of Pylance.

**Comment by smeden-lod** [comment]
Hi, I followed [Luciano Ramalho's talk about typing max function](https://www.youtube.com/watch?v=kDDCKwP7QgQ)  using Protocol and bounded TypeVar :

```python
class Orderable(Protocol):

    def __lt__(self, other: Any): ...


S = TypeVar('S', bound=Orderable)

def t_h(xs: List[S]) -> Optional[S]:
    if len(xs) == 0:
        return None
    else:
        m = xs[0]
        for x in xs:
            if m < x:
                m = x
        return m


#assert t_h([]) is None
# assert t_h([7, 11, 42, 5, 1]) == 42
# assert t_h([1.2, .628, math.pi, math.e, .1]) == math.pi
assert t_h(["python", "Haskell", "C", "TypeScript", "ELM", "Clojure"]) == max(
    ["python", "Haskell", "C", "TypeScript", "ELM", "Clojure"])
print("tests ok")
```
pyright 1.1.169 complains that 

```
Argument of type "list[str]" cannot be assigned to parameter "xs" of type "List[S@t_h]" in function "t_h"
    TypeVar "_T@list" is invariant
      Type "Orderable" cannot be assigned to type "str"
```
when mypy 0.812 won't. I might miss something, I thought the issue resembles the above mentioned one. Any hint ?

**Comment by smeden-lod** [comment]
I understood it might have something to do with `List[S]` that turns `S` into some invariant type...

**Comment by erictraut** [comment]
The problem is that the `__lt__` method you've declared does not include a return type annotation. It therefore isn't compatible with the `__lt__` method on the `str` class. If you add a `-> bool` to the `__lt__` method, the code will type check without errors.

**Comment by smeden-lod** [comment]
Thanks for such a quick answer !

Stéphane

Le dim. 28 nov. 2021 à 11:57, Eric Traut ***@***.***> a
écrit :
